src/pudl/transform/ferc1.py

```diff
@@ -2731,6 +2731,24 @@ def remove_nones_in_list(list_of_things):
             tbl_meta.loc[xbrl_factoid, "calculations"] = json.dumps(calc_to_update)
         return tbl_meta.reset_index()
 
+    def process_xbrl_metadata_calculations(
+        self,
+    ) -> pd.DataFrame:
+        """Convert xbrl metadata calculations into a table of calculation components."""
+        metadata = self.xbrl_metadata
+        if all(metadata.calculations == "[]"):
+            logger.info(f"{self.table_id.value}: No calculations.")
+            return pd.DataFrame()
+
+        calc_dfs = []
+        for calc, factoid in zip(metadata.calculations, metadata.xbrl_factoid):
+            calc_dfs.append(
+                pd.DataFrame(json.loads(calc)).assign(
+                    table_name=self.table_id.value, xbrl_factoid=factoid
+                )
+            )
+        return pd.concat(calc_dfs).reset_index(drop=True).explode("source_tables")
```

Comment:
I've also found myself using this anti-pattern. I feel like there should be a simple way to do this that doesn't involve iterating over the rows. Like using `df.apply()` and/or `pd.json_normalize()`?

If you do need to iterate over a dataframe, rather than zipping columns together there's also the built-in [`df.itertuples()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.itertuples.html) which lets you refer to named elements in a single row like:

```py
for row in metadata.itertuples():
    calc_dfs.append(
        pd.DataFrame(json.loads(row.calculations)).assign(
            table_name=self.table_id.value, xbrl_factoid=row.xbrl_factoid
        )
    )
```

Reply:
vv good point! i made a big revision here